addons/pos_loyalty/static/src/js/Loyalty.js

```diff
@@ -426,10 +529,47 @@ const PosLoyaltyOrder = (Order) => class PosLoyaltyOrder extends Order {
                 await this.pos.fetchCoupons([['id', 'in', couponsToFetch]], couponsToFetch.length);
                 // Remove coupons that could not be loaded from the db
                 this.codeActivatedCoupons = this.codeActivatedCoupons.filter((coupon) => this.pos.couponCache[coupon.id]);
-                this.couponPointChanges = Object.fromEntries(Object.entries(this.couponPointChanges).filter((k, pe) => this.pos.couponCache[pe.coupon_id]));
+                this.couponPointChanges = Object.fromEntries(Object.entries(this.couponPointChanges).filter(([k, pe]) => this.pos.couponCache[pe.coupon_id]));
             }
         });
     }
+    /**
+     * The idea is to update the quantity of the reward line when the coupon points decreased.
+     * We can identify that there is a decrease in coupon points when the remaining points for
+     * the coupon becomes negative. Out of the negative points, we recompute the quantity that
+     * will make the coupon points to become zero.
+     *
+     * E.g.
+     *
+     * Current order:
+     * > 10 product1 --(producing)-> 10 points
+     * > 4 product2 (reward line) (2 points consumed per item) --(consuming)-> 8 points
+     * > Remaining points: 2
+     *
+     * User action:
+     * > 10 product1 --(reduced to)-> 4 product1 --(producing only)-> 4 points
+     *
+     * Result of user action:
+     * > coupon's remaining points: 4 - 8 = -4
+     *
+     * Correct negative remaining points:
+     * > Automatically set quantity of reward line to 2 (4 points consumed)
+     * > Remaining points becomes: 0
+     *
+     * @param {Orderline} rewardLine
+     * @return {number} The maximum possible quantity for the given reward line when the remaining points of the coupon is negative.
+     */
+    _getNewRewardQuantity(rewardLine) {
+        const remainingPoints = rewardLine.order._getRealCouponPoints(rewardLine.coupon_id, true);
+        let correction = 0;
+        if (remainingPoints < 0) {
+            // We correct the quantity of this line if the remaining points is negative.
+            const reward = rewardLine.order.pos.reward_by_id[rewardLine.reward_id];
```

Comment:
you're inside the concerned order
```suggestion
        const remainingPoints = this._getRealCouponPoints(rewardLine.coupon_id, true);
        let correction = 0;
        if (remainingPoints < 0) {
            // We correct the quantity of this line if the remaining points is negative.
            const reward = this.pos.reward_by_id[rewardLine.reward_id];
```